web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/tracecontext/Tracestate.java

```diff
@@ -11,27 +15,97 @@
  *
  * Implementations can add vendor specific details here.
  */
-@Experimental
 public class Tracestate {
 
     /**
-     * Field to capture tracestate header
+     * Internal representation of the tracestate
+     */
+    private LinkedHashMap<String, String> internalList = new LinkedHashMap<String, String>(32);
+
+    /**
+     * String representation of the tracestate
      */
-    public final String reserved;
+    private String internalString = null;
+
+    private static String KEY_WITHOUT_VENDOR_FORMAT = "[a-z][_0-9a-z\\-\\*\\/]{0,255}";
+    private static String KEY_WITH_VENDOR_FORMAT = "[a-z][_0-9a-z\\-\\*\\/]{0,240}@[a-z][_0-9a-z\\-\\*\\/]{0,13}";
+    private static String KEY_FORMAT = KEY_WITHOUT_VENDOR_FORMAT + "|" + KEY_WITH_VENDOR_FORMAT;
+    private static String VALUE_FORMAT = "[\\x20-\\x2b\\x2d-\\x3c\\x3e-\\x7e]{0,255}[\\x21-\\x2b\\x2d-\\x3c\\x3e-\\x7e]";
+
+    private static Pattern KEY_VALIDATION_RE = Pattern.compile("^" + KEY_FORMAT + "$");
+    private static Pattern VALUE_VALIDATION_RE = Pattern.compile("^" + VALUE_FORMAT + "$");
+
+    private static String DELIMITER_FORMAT = "[ \\t]*,[ \\t]*";
+    private static String MEMBER_FORMAT = String.format("(%s)(=)(%s)", KEY_FORMAT, VALUE_FORMAT);
+
+    private static Pattern DELIMITER_FORMAT_RE = Pattern.compile(DELIMITER_FORMAT);
+    private static Pattern MEMBER_FORMAT_RE = Pattern.compile("^" + MEMBER_FORMAT + "$");
 
     /**
      * Ctor that creates tracestate object from given value
      */
-    public Tracestate(String value) {
-        if (value == null || value.length() == 0) {
-            throw new IllegalArgumentException("invalid spanId");
+    public Tracestate(String input) {
+        if (input == null) {
+            throw new IllegalArgumentException("input is null");
+        }
+        
+        String[] values = DELIMITER_FORMAT_RE.split(input);
+        for (String item : values) {
+            Matcher m = MEMBER_FORMAT_RE.matcher(item);
+            if (!m.find()) {
+                throw new IllegalArgumentException(String.format("invalid string %s in tracestate", item));
+            }
+            String key = m.group(1);
+            String value = m.group(3);
+            if (internalList.get(key) != null) {
+                throw new IllegalArgumentException(String.format("duplicated keys %s in tracestate", key));
+            }
+            internalList.put(key, value);
         }
-        reserved = value;
+        if (internalList.size() > 32) {
+            throw new IllegalArgumentException("cannot have more than 32 key-value pairs");
+        }
+        internalString = toInternalString();
     }
 
+    /**
+     * Ctor that creates a tracestate object from a parent one
+     */
+    public Tracestate(Tracestate parent, String key, String value) {
```

Comment:
For application insights case we need another helper method also - which can basically swap the value of the `azure key`. Should this helper reside here? If not we would need to have almost similar method in the AI implementation.

Reply:
Can we use ```new Tracestate(tracestate, "key", "new value")```?

Reply:
should work